### ISSUE TITLE
Kafka postinstallation error: symlink exists '/opt/bitnami/kafka/conf' -> '/opt/bitnami/kafka/config'

### ISSUE THREAD

**Issue by tgrzesik** [bitnami/bitnami-docker-kafka#45]
Hey,
I'm trying to run kafka cluster with bitnami helm chart (https://github.com/bitnami/charts/tree/master/bitnami/kafka) on bare-metal kubernetes with the following errors at runtime:
```
kafka   INFO  ==> No injected configuration files found. Creating default config files...
Error executing 'postInstallation': EEXIST: file already exists, symlink '/opt/bitnami/kafka/conf' -> '/opt/bitnami/kafka/config'
```
I have tried with many tags (2.2.0, 2.0.0, 2.2.2-debian-r0, 1.1.1 etc.) with no result.

I found similar errors with apache (link to the issue: https://github.com/bitnami/bitnami-docker-opencart/issues/37). I'm using nfs-client persistencevolume type on my k8s cluster. Is it the same issue? It will be pretty hard to convince my team to change storage type just to make kafka work ;)

Any solutions or workarounds?

Thanks!

**Comment by marcosbc** [comment]
Hi @tgrzesik, it seems like you are not persisting the right directory or file.

You should be persisting `/bitnami/kafka`, but it seems like you might be persisting `/opt/bitnami/kafka/config`. Could you check if this is the case?

Also, if you're mounting any configuration file, you should be doing it from the directory `/opt/bitnami/kafka/conf` instead of `/opt/bitnami/kafka/config`.

**Comment by tgrzesik** [comment]
@marcosbc thanks for your fast reply.
I'm running on default statefulset config from helm chart. I don't think that's the issue especially when the same configuration on another k8s cluster works like a charm.

**Comment by marcosbc** [comment]
Hi @tgrzesik, unfortunately we don't know what could be happening in your case. Would you happen to know if the NFS disk is only being used for the persisted data volumes (`/bitnami`)?

The implementation of the Kafka initialization copies the contents of `config` to `conf`, then deletes the directory and finally creates a symlink (`config -> conf`).

Therefore, either `config` is not being deleted properly, or a configuration file was mounted in that directory.

Could you launch a new clean Helm chart with `image.debug` set to `true`? That should provide more info for debugging the issue on our side. Also, if you could provide your `values.yaml` file (hiding sensitive info) it would be great.

**Comment by jaskirat8** [comment]
I also faced same issue. But interesting thing is it worked on latest docker but for my server having docker Server Version: 17.09.0-ce this issue poped up.
I am not using any kind of data persistence and just raw same config(I am sure because one is staging and another is QA server for same code controlled by CI system)

Below is the compose file snapshot  :
```
services:
  zookeeper:
    deploy:
      mode: replicated
      replicas: 1
    environment:
      TZ: Asia/Kolkata
    image: zookeeper:3.5.5
  kafka:
    deploy:
      mode: replicated
      replicas: 1
    environment:
      ALLOW_PLAINTEXT_LISTENER: "yes"
      KAFKA_CFG_BROKER_ID: '1'
      KAFKA_CFG_HOST_NAME: kafka
      KAFKA_CFG_ZOOKEEPER_CONNECT: zookeeper:2181
      TZ: Asia/Kolkata
    image: bitnami/kafka:2.2.1
```

**Comment by marcosbc** [comment]
Hi @jaskirat8, we were able to get it working with the following `docker-compose.yml` file:

```
version: '2'

services:
  zookeeper:
    image: zookeeper:3.5.5
  kafka:
    environment:
      - ALLOW_PLAINTEXT_LISTENER=yes
      - KAFKA_CFG_BROKER_ID=1
      - KAFKA_CFG_HOST_NAME=kafka
      - KAFKA_CFG_ZOOKEEPER_CONNECT=zookeeper:2181
    image: bitnami/kafka:2.2.1
```

Note that we are using the latest release of `bitnami/kafka:2.2.1` which is `8f763c50` (or `bitnami/kafka:2.2.1-debian-9-r22`).

In recent releases we changed the approach of how the Kafka container works. Could you verify your Kafka image is updated?

**Comment by jaskirat8** [comment]
@marcosbc Thanks. But as you can refer my snapshot we are using the same imahe as you suggested and got same error.

Basically 2.2.1 message was similar in nature but wording was like "rm /pathtofile File not found" 

When i downgraded image to 2.2.0 the error message was same as this issues heading.


Its just a single node swarm setup with local setup and rhel/centos 7 in both setups only docker version is different as i mentioned.

You need any extra info let me know.

**Comment by marcosbc** [comment]
Hi @jaskirat8, we're interested in knowing the exact error you got in 2.2.1 (latest tag), as we made a major change to the container logic (although from an external point of view it should be the same). Could you share this with us?

In any case, if the error is related to removing the configuration directory, it is possible it was mounted as a volume somehow, and both errors are similar. Could you check if you have mounted any configuration file manually, or the configuration directory? If so that could be the reason for the issue in both image versions.

**Comment by jaskirat8** [comment]
I will get the exact message snapped but yes it says error removing configuration directory. And no volume mount as you can see the compose. No volume mounts. I understand the logic and tried searching for mounts only but even docker service inspect didn't give any clue of volume.

To be clear of any dangaling issues, it reset docker by deleting /var/lib/docker even then exact same error. 

You can see there is no scope of file being there as its purely container created namespace and volume. 

The change you are talking about can you check the permissions being given to existing directory path and the shell executing user is mot causing issues ?

Any extra information required from stack or docker logs, do let me know.

**Comment by santi698** [comment]
Hi! I'm facing a similar issue when starting a previously stopped container.
The error occurs when [this line](https://github.com/bitnami/bitnami-docker-kafka/blob/master/2/debian-9/rootfs/libkafka.sh#L314) is executed. The command exits with code 1 and prints:
`cp: '/opt/bitnami/kafka/config/.' and '/opt/bitnami/kafka/conf/.' are the same file`.
Inspecting the container's file system I found that `/opt/bitnami/kafka/config` is a symlink to `/opt/bitnami/kafka/conf` as you mentioned, and trying to cp from conf into config gives exit code 1 too.
Hope this helps

**Comment by marcosbc** [comment]
Hi @santi698, thanks for the details you provided. Because of it we were able to reproduce the issue by following these steps:

- Start the docker-compose deployment
- Stopping one container with `docker stop` (not `docker-compose down`)
- Starting that container again with `docker start`

We are working on fixing this issue. We expect to have the fix released tomorrow.

In the meantime, if you are using Docker Compose for your deployment you can do the following:

```
docker-compose down
docker-compose up
```

**Comment by marcosbc** [comment]
Hi, this issue has been fixed now in `1.1.1-debian-9-r237`, `2.2.1-debian-9-r27`, `1.1.1-ol-7-r319` and `2.2.1-ol-7-r27`. Let us know if you keep finding issues!
